### ISSUE TITLE
Elasticsearch/templating: Values in variables from Elasticsearch datasources never get refreshed

### ISSUE THREAD

**Issue by slhck** [grafana/grafana#60102]
**What I want to achieve:**

I have an Elasticsearch data source from which I fetch the values of a variable. For instance, this could be a list of server hostnames. The query in the variable settings might be:

```
{"find": "terms", "field": "server.hostname.keyword" }
```

The variable has a "Include all option" set, and multi-value is checked. The variable is set to refresh on time range change.

I want to ensure that when new values appear in this variable, users can select it in the dashboard.

**What happened**:

Now, when new hosts appear in the Elasticsearch index, the values of this variable do not change. The list of available hosts stays the same.

This is even the case when I do change the time range – which, I assumed, would reload the values.

**What you expected to happen**:

I would expect the list of values to include the new host – at least when I change the time range.

Instead, I have to actually do the following:

* Go to the dashboard variables
* Clicking "Run Query" does not show the new value!
* Select the same Elasticsearch index data source again
* Press "Run Query"
* Now the new values appear in the list
* Click apply and I can select them in the dashboard

Even when I set those variables to reload on dashboard load, simply doing that does **not** reload the values. I have to go through the above steps.

**How to reproduce it (as minimally and precisely as possible)**:

* Set up an Elasticsearch data source
* Index a few documents with a keyword type
* Set up the dashboard to use a variable based on querying that keyword (see query above) and make it reload on time range change
* Open the dashboard
* Index new data with a new keyword value into the Elasticsearch index
* Reload the dashboard or change the time range
* See that the value does not appear in the list

**Anything else we need to know?**:

This is the templating part of the dashboard:

```json
{
  "templating": {
    "list": [
      {
        "current": {
          "selected": true,
          "text": [
            "All"
          ],
          "value": [
            "$__all"
          ]
        },
        "datasource": {
          "type": "elasticsearch",
          "uid": "3GNKgHVVk"
        },
        "definition": "{\"find\": \"terms\", \"field\": \"server.hostname.keyword\" }",
        "hide": 0,
        "includeAll": true,
        "multi": true,
        "name": "client",
        "options": [],
        "query": "{\"find\": \"terms\", \"field\": \"server.hostname.keyword\" }",
        "refresh": 2,
        "regex": "",
        "skipUrlSync": false,
        "sort": 1,
        "type": "query"
      }
    ]
  }
}
```

**Environment**:

- Grafana version: 9.3.1
- Data source type & version: Elasticsearch 8.4.2 or higher
- OS Grafana is installed on: Docker
- User OS & Browser: Current Chrome, OS is irrelevant
- Grafana plugins: None
- Others: None

**Comment by gabor** [comment]
@slhck thanks for reporting the problem.

i tried it and i cannot reproduce it, for me it behaves as it should, when new data is inserted into elastic, and i change the time-range, it shows the new values.

i think we can focus on the part where you directly go to the variable-settings and press "run query" and the new values do not show up.

the question is, is an elastic-request sent at that point or not?

could you dig a little using the browser tools?:

1. open the developer-tools in chrome (view/developer/developer_tools), choose the network tab.
2. in grafana, click on "run query". 
3. does a new entry appear in the network tab in the developer tools?

on my computer, a new entry with the name `_msearch` appears here. i can click on it, choose the "Preview" tab, and i can investigate the data returned from elastic, and i can confirm that it contains the new values.

could you try this please?

**Comment by slhck** [comment]
Thanks for your quick feedback! I think what may have caused this issue is that this dashboard came from a different instance of Grafana, from which I took the dashboard JSON representation, exchanged the data source UIDs, and pasted it into a new instance.

I tried to reproduce this via these steps:

- Open another Grafana instance (same version 9.3.1)
- Copy the JSON from a dashboard using variables pointing to an Elasticsearch data source UID (e.g. `QZNFgN44k`)
- The JSON contains the templating I showed in the original issue description (i.e., an Elasticsearch field keyword query)
- Replace the UID in the JSON with a datasource UID from the new instance (e.g. `e7MDJ754k`)
- On the instance, go to Dashboards and click Import, then paste the JSON into the field

But for some reason, it worked. I will close this issue for now but when I encounter it again, I will make sure to check the network requests to debug it.
